language-server/src/references.js

```diff
@@ -0,0 +1,174 @@
+import { fetchDocument } from "./documents.js";
+import { getEmbeddedIdentifer, getIdentifier } from "./identifiers.js";
+import { keywordNameFor } from "./json-schema.js";
+import { JsoncInstance } from "./jsonc-instance.js";
+import { getSemanticTokens } from "./semantic-tokens.js";
+import { getSchemaResources, workspaceUri } from "./server.js";
+import { isAbsoluteUrl, isAnchor } from "./util.js";
+import { dirname, join } from "node:path";
+import { pathToFileURL, resolve } from "node:url";
+
+/**
+ * `documentUri` -> `ref -> instance`
+ * @type {Map<string, Map<string, import("./jsonc-instance.js").JsoncInstance>}
+ */
+export const references = new Map();
+
+/**
+ * @param {string} uri
+ */
+export const getReferenceForDocument = (uri) => {
+  return references.get(uri);
+};
+
+export const deleteReferencesForDocument = (uri) => {
+  references.delete(uri);
+};
+
+/**
+ * @param {import("./jsonc-instance.js").JsoncInstance} keywordInstance
+ * @param {string} dialectUri
+ */
+export const addReference = (keywordInstance, dialectUri) => {
+  const referenceKeywordIds = [
+    "https://json-schema.org/keyword/ref",
+    "https://json-schema.org/keyword/draft-04/ref"
+  ];
+  const referenceKeywordNames = referenceKeywordIds.map((keywordId) => keywordNameFor(keywordId, dialectUri));
+  if (!referenceKeywordNames.includes(keywordInstance.value()) || keywordInstance?.node?.parent?.children?.length < 2) {
+    return;
+  }
+  const { pointer, textDocument, node } = keywordInstance;
+
+  const valueInstance = keywordInstance._fromNode(node.parent.children[1], keywordInstance.pointer);
```

Comment:
The point of the `JsoncInstance` class is to abstract the underlying AST being used. We're currently using `jsonc-parser`'s AST, but that will probably need to change at some point and ideally, all we should have to do is to replace `JsoncInstance` with a new `*Instance` class and everything should still work.

All that is to say, don't use `node` outside of `JsoncInstance`. I should have made it private, but support for private members in classes is a relatively new feature and I forgot it existed. If you need to traverse the AST, use the utility functions provide in `JsoncInstance` or make a new one if what you need isn't available.

Also, `_fromNode` was intended to be private as well.